src/metabase/models/recent_views.clj

```diff
@@ -79,18 +121,177 @@
+    [:timestamp :string]]
+   [:multi {:dispatch :model}
+    [:card [:map
+            [:parent_collection ::pc]
+            [:display :string]
+            [:moderated_status [:enum "verified" nil]]]]
+    [:dataset [:map
+               [:parent_collection ::pc]
+               [:moderated_status [:enum "verified" nil]]]]
+    [:dashboard [:map [:parent_collection ::pc]]]
+    [:table [:map
+             [:display_name :string]
+             [:database [:map
+                         [:id [:int {:min 1}]]
+                         [:name :string]]]]]
+    [:collection [:map
+                  [:parent_collection ::pc]
+                  [:authority_level [:enum :official nil]]]]]])
+
+(defmulti fill-recent-view-info
+  "Fills in additional information for a recent view, such as the display name of the object.
+
+  - When called from `GET /popular_items`, the `model_object` field will be present, and should be used instead of
+  querying the database for the object."
+  (fn [{:keys [model #_model_id #_timestamp card_type]}]
+    (or (get {"model" :dataset
+              "question" :card} card_type)
+        (keyword model))))
+
+(defmethod fill-recent-view-info :default [m] (throw (ex-info "Unknown model" {:model m})))
+
+(defn get-parent-coll
+  "Gets parent collection info for a recent view item."
+  [coll-id-or-coll]
+  (-> (cond (map? coll-id-or-coll) (if-let [parent-id (:parent_id (t2/hydrate coll-id-or-coll :parent_id))]
+                                     ;; hydrate the effective location on the collection
+                                     (t2/select-one :model/Collection parent-id)
+                                     (root/root-collection-with-ui-details {}))
+            (nil? coll-id-or-coll) (root/root-collection-with-ui-details {})
+            :else (t2/select-one :model/Collection coll-id-or-coll))
+      (select-keys [:id :name :authority_level])
+      (update :authority_level #(some-> % keyword))))
+
+(mu/defn get-moderated-status
+  "Returns moderated_status for a given model and model-id.
+
+  (Currently only used for cards and models, but ought to be extended to dashboards in the future)"
+  [model :- [:enum :card] model-id] :- [:maybe "verified"]
+  (-> (t2/select-one [:model/ModerationReview :status]
+                     {:where [:and
+                              [:= :moderated_item_id model-id]
+                              [:= :moderated_item_type (name model)]
+                              [:= :most_recent true]]})
+      :status))
+
+(defn ellide-archived
+  "Returns nil if the model is archived, otherwise returns the model.
+  We use this to ensure that archived models are not returned in the recent views."
+  [model]
+  (cond
+    (false? (:archived model)) model
+
+    (true? (:archived model)) nil
+
+    (nil? (:archived model))
+    (throw (ex-info "Archived field is nil" {:model model}))))
+
+(defmethod fill-recent-view-info :card [{:keys [_model model_id timestamp model_object]}]
```

Comment:
Would it make sense for the impls of this multimethod to live in the model namespaces for each model? That way each model is responsible for what information is included in recent views, rather than blowing up this namespace with all the implementations. That's the pattern I followed with `audit-log/model-details` for instance.

Don't need to block on this, just a suggestion but it could go either way.

Reply:
Glad you asked! I thought about that too, but I figured that I wanted these methods to be seen from the POV of a recent view item, instead of a model. Part of the reason for that is the impedence mismatch between :model/Card being both questions and models (and metrics etc if we ever add those?). So I thought it was more simple to handle things as they are on the api side, instead of the way they are on our model/instance side.